### PR TITLE
thread_local: be excruciatingly explicit in dtor code

### LINKED ISSUE #124317: The implementation in `fast_local.rs` seems to violate pointer aliasing rule
https://github.com/rust-lang/rust/blob/c1feb3eceef7d5f0126c309a87062cf413fe0a25/library/std/src/sys/thread_local/fast_local.rs#L210
````rust
unsafe fn try_register_dtor(&self) -> bool {
        match self.dtor_state.get() {
            DtorState::Unregistered => {
                // SAFETY: dtor registration happens before initialization.
                // Passing `self` as a pointer while using `destroy_value<T>`
                // is safe because the function will build a pointer to a
                // Key<T>, which is the type of self and so find the correct
                // size.
                unsafe { register_dtor(self as *const _ as *mut u8, destroy_value::<T>) };  // #1
                self.dtor_state.set(DtorState::Registered);
                true
            }
            DtorState::Registered => {
                // recursively initialized
                true
            }
            DtorState::RunningOrHasRun => false,
        }
    }
````
At `#1`, acquire a pointer of type `* mut u8` from a shared reference `&Key<T>`, and subsequently, in the `destroy_value`, we reinterpret `* mut u8` as `* mut Key<T>` and use it as a mutable pointer https://github.com/rust-lang/rust/blob/c1feb3eceef7d5f0126c309a87062cf413fe0a25/library/std/src/sys/thread_local/fast_local.rs#L240
````rust
unsafe extern "C" fn destroy_value<T>(ptr: *mut u8) {
    let ptr = ptr as *mut Key<T>;

    // SAFETY:
    //
    // The pointer `ptr` has been built just above and comes from
    // `try_register_dtor` where it is originally a Key<T> coming from `self`,
    // making it non-NUL and of the correct type.
    //
    // Right before we run the user destructor be sure to set the
    // `Option<T>` to `None`, and `dtor_state` to `RunningOrHasRun`. This
    // causes future calls to `get` to run `try_initialize_drop` again,
    // which will now fail, and return `None`.
    //
    // Wrap the call in a catch to ensure unwinding is caught in the event
    // a panic takes place in a destructor.
    if let Err(_) = panic::catch_unwind(panic::AssertUnwindSafe(|| unsafe {
        let value = (*ptr).inner.take();  // #2
        (*ptr).dtor_state.set(DtorState::RunningOrHasRun);
        drop(value);
    })) {
        rtabort!("thread local panicked on drop");
    }
}
````
At `#2` the method `take` of the field `inner` requires `& mut self` as defined in https://github.com/rust-lang/rust/blob/c1feb3eceef7d5f0126c309a87062cf413fe0a25/library/std/src/sys/thread_local/mod.rs#L98

The whole process can be that we use `&Key<T>` to acquire a `& mut Key<T>` and call a mutable method on that variable, which seems problematic.

### PR DESCRIPTION
**Pull request by workingjubilee**
Use raw pointers to accomplish internal mutability, and clearly split references where applicable. This reduces the likelihood that any of these parts are misunderstood, either by humans or the compiler's optimizations.

Fixes #124317

r? @joboet

### CODE REVIEW

File: library/std/src/sys/thread_local/fast_local.rs [hunk 1]
```diff
@@ -237,8 +237,9 @@ unsafe extern "C" fn destroy_value<T>(ptr: *mut u8) {
     // Wrap the call in a catch to ensure unwinding is caught in the event
     // a panic takes place in a destructor.
     if let Err(_) = panic::catch_unwind(panic::AssertUnwindSafe(|| unsafe {
-        let value = (*ptr).inner.take();
-        (*ptr).dtor_state.set(DtorState::RunningOrHasRun);
+        let Key { inner, dtor_state } = &*ptr;
+        let value = inner.take();
+        dtor_state.set(DtorState::RunningOrHasRun);
         drop(value);
```
Comment (workingjubilee) [inline library/std/src/sys/thread_local/fast_local.rs:243]: something that's not clear to me is whether we can simply rearrange this: do `dtor_state.set(DtorState::RunningOrHasRun)` and then call the more lightweight `drop_in_place`?

File: library/std/src/sys/thread_local/fast_local.rs [hunk 2]
```diff
@@ -237,8 +237,9 @@ unsafe extern "C" fn destroy_value<T>(ptr: *mut u8) {
     // Wrap the call in a catch to ensure unwinding is caught in the event
     // a panic takes place in a destructor.
     if let Err(_) = panic::catch_unwind(panic::AssertUnwindSafe(|| unsafe {
-        let value = (*ptr).inner.take();
-        (*ptr).dtor_state.set(DtorState::RunningOrHasRun);
+        let Key { inner, dtor_state } = &*ptr;
+        let value = inner.take();
```
Comment (workingjubilee) [inline library/std/src/sys/thread_local/fast_local.rs:241]: The purpose of adding this two-step is to make it clear, combined with the other change, that we are using internal mutability and not relying on upgrading the pointer to `&mut`.

File: library/std/src/sys/thread_local/mod.rs [hunk 3]
```diff
@@ -91,13 +91,15 @@ mod lazy {
...
-        /// As such, callers of this method must ensure no `&` and `&mut` are
-        /// available and used at the same time.
+        /// Watch out: unsynchronized internal mutability!
+        ///
+        /// # Safety
+        /// Unsound if called while any `&'static T` is active.
         #[allow(unused)]
-        pub unsafe fn take(&mut self) -> Option<T> {
-            // SAFETY: See doc comment for this method.
-            unsafe { (*self.inner.get()).take() }
```
Comment (workingjubilee) [inline library/std/src/sys/thread_local/mod.rs:100]: This previous version of this `fn take` was actually perfectly sound **and safe** in isolation... if we have `&mut self`, we should be able to take `Option<T>`, of course... but combined with all the pointer casting it is an attractive nuisance.

File: library/std/src/sys/thread_local/fast_local.rs (hunk 1 above)
Comment (joboet) [reply library/std/src/sys/thread_local/fast_local.rs:243]: We could, and there are some other improvements possible as well, see #116123.

File: library/std/src/sys/thread_local/mod.rs [hunk 4]
```diff
@@ -91,13 +91,15 @@ mod lazy {
             }
         }
 
-        /// The other methods hand out references while taking &self.
-        /// As such, callers of this method must ensure no `&` and `&mut` are
-        /// available and used at the same time.
+        /// Watch out: unsynchronized internal mutability!
+        ///
+        /// # Safety
+        /// Unsound if called while any `&'static T` is active.
```
Comment (joboet) [inline library/std/src/sys/thread_local/mod.rs:97]: References may be active upon entry here, but they mustn't be used anymore.
```suggestion
        /// Causes UB if any reference to the value is used after this.
```

**Review by joboet — APPROVED** [review]
👍 

r=me with the updated safety comment

**Pull request merged**